ENCM-84 Make a matrix for Degron experiments

forresttanaka opened the pull request:
This is one of the simpler matrices in that it mostly follows what other matrices do. Even the basic Experiment matrix is more complicated because it has a nested Y and X axes where this one has only a nested X-axis. The important part is in matrix_degron.js. A lot of the other changed files generate data to test the Degron matrix locally.

forresttanaka committed:
ENCM-84-degron-matrix

* Add Degron matrix path and query to the experiment object.
* Generates and displays the Degron matrix header row. Generates but doesn’t yet display the data rows.
* Displaying complete data in table.
* Display the data rows. Make the basic styling for a horizontally scrollable matrix work.
* Remove the bold header titles.
* Calculate the proper number of columns in the matrix. Change the link to the degron matrix page to include a filter for GM name so that we don’t grab a bunch of experiments with no GM.
* Update the card link so that it goes to more manageable data. Add another button to link to the degron matrix with data that works locally.
* Try filling unused cells with an empty div instead of null to see if that will fill out the table instead of leaving trailing cells non-existent.
* Not accounting for the header when counting row cells.
* Style the table cell borders more consistently.
* Change cells to have a colored block instead of a count.
* Add links to the column headers.
* Link cells to a corresponding search. Transform ChIP-seq titles to strip prefixes as the target label display makes those prefixes redundant.
* Sort table columns in a predefined order.
* Remove Clear All Selections button which will make a GIGANTIC matrix that probably won’t load.
* Don’t show the Visualize button.
* Restyle search form
* Update pre-defined sort order for matrix columns.
* Correct column sorting. A couple of the pre-defined sorting keys were typos, and I was using sortBy as if we used Lodash instead of Underscore.
* Increase row padding.
* Add ARIA labels to each cell.
* Remove all scrolling code that for some reason hasn’t had any effect in probably years, and no one complained.
* Move MatrixAddCart to matrix.js so that anyone can use it.
* Refactor MatrixAddCart to not require so many properties, and to request far less data from the server. Not sure it works as not enough local data exists, so will test on a demo.
* Add tje All All to Cart button to Degron matrix.
* Correct the CSS from last commit.
* Further corrections to CSS in last commit.
* Make main navigation menus not auto dropdown. Add a collection menu item for the Degron matrix. Make the menu fonts smaller to try to make them fit smaller screens better.
* Remove test file.
* Add backend test for the Degron matrix.
* Correct search-views test syntax error.
* Add BDD tests for Degron matrix.
* Make header row fixed height so that the links within them all have the same height.
* Change the title of the Degron home-page card and matrix page title according to the ticket. Also add a home-page search string, though that needs to change once we have the final URL for the Degron page.
* Change the search views test for the new title for the Degron matrix.
* Add Khine’s Degron icon for the home-page button.
* Change the home-page button and search links for Degron to their final form, using the Degron internal tag.
* Change all the links and search strings to use the Degron internal tag.
* Revert my change for the auto dropdown menus. Change the Degron matrix BDD test to work with the new menu.
* Add test data inserts for Degron experiments.
* Update BDD test for new Degron experiment data.
* Try fixes for indexing and BDD tests with the new test inserts.
* Try to fix BDD test failures. The last commit fixed the indexing test failure.
* Try to fix BDD test failures.
* Try to get CircleCI BDD test to pass.

dougli1sqrd reviewed (commented):
The first thing I'm noticing is when I very first go to the Degron page the horizontal items are all squished together: Home page -> Data drop down to Degron link -> squished.

<img width="1044" alt="Screen Shot 2022-11-09 at 1 45 03 PM" src="https://user-images.githubusercontent.com/1422171/200948436-48c7adfd-ebb7-4fdc-8ffb-6915846357c5.png">

After a refresh on the page it behaves normally. But I can always reproduce by navigating away, and navigating back to the page (even with just forward and back buttons).
<img width="809" alt="Screen Shot 2022-11-09 at 1 46 33 PM" src="https://user-images.githubusercontent.com/1422171/200948659-92e386ac-dcbb-435f-b532-2b0bcaee40a4.png">

forresttanaka commented:
Oh, that’s a Safari thing apparently, which I forgot to test. I’ll need to fix that. Hopefully it’s not difficult.

forresttanaka commented:
Looks like Firefox has a different issue.

dougli1sqrd commented:
Should I also see cells filled in at all in that view that I screenshotted?

forresttanaka commented:
Yes, like it appears in Chrome.

dougli1sqrd commented:
ah yeah, looks like something is up in firefox then there

forresttanaka commented:
I have fixes for Safari and Firefox ready, so I’ll push those and whatever else we need once you’re done with code review.

dougli1sqrd commented on src/encoded/static/components/matrix_degron.js:
```diff
@@ -0,0 +1,403 @@
+ * @param {object} colSubcategoryBucket Bucket for the current subcolumn
+ * @returns {string} Query string for the column header
+ */
+const composeColCategoryQuery = (context, colCategoryBucket, colSubcategoryBucket) => {
```
How come we say "context" for the search result object? Would another name that just describes the fact that it's the search results be more clear? Or is there a historical reason here?

forresttanaka replied:
“context” refers to any object from the back end. It got that name before I started and by that time it had already become ubiquitous through the code base. Laurence felt some regret calling it that once React also got the concept of a “context.”

dougli1sqrd replied:
ahh yeah okay. Yeah I did think at first it could be related to the react context, but we have that also.

dougli1sqrd commented on src/encoded/static/components/matrix_degron.js:
```diff
@@ -0,0 +1,403 @@
+/**
+ * Defines the sorting order of the assay/target table columns.
+ */
+const COL_CATEGORY_DATA_SORT_ORDER = [
```
Why is this the sorting order? Maybe we should comment also how we know this so in case something has to change we know how we generated this order in the first place.

forresttanaka replied:
Ben defined the sorting order. I don’t know why it’s that order.

dougli1sqrd replied:
okay!

dougli1sqrd reviewed (approved):
Happy to take a look at your upcoming updates

forresttanaka commented:
Could you reject this PR ? Then I can put in the updates and re-request your review. _Edit: or “Request changes” rather_.

dougli1sqrd reviewed (changes requested):
Approve for what's currently here, awaiting updates from earlier convo

forresttanaka committed:
ENCM-84-degron-matrix

Fix the Degron matrix appearance on Safari and Firefox where all the columns were squished into one.

dougli1sqrd commented:
does the demo get updated? Or should I try running this locally? I guess that's what QA is for too though 🤷

forresttanaka commented:
I updated the demo and it’s rebuilding right now. I’ll wait for the tests to pass before putting it back into code review.

dougli1sqrd reviewed (dismissed):
@forresttanaka this worked!

forresttanaka committed:
ENCM-84-degron-matrix

Have home-page Degron button highlight for relevant home-page searches. Make lighter boxes for Degron matches with only one treatment.

forresttanaka committed:
ENCM-84-degron-matrix

Correction to the CSS in the last commit.

forresttanaka committed:
ENCM-84-degron-matrix

Update BDD tests to fix the failure after adding the home-page Degron button to the search-result highlighting.

forresttanaka committed:
ENCM-84-degron-matrix

Further updates to the home-page BDD test.
